keps/sig-node/20190717-seccomp-ga.md

```diff
@@ -0,0 +1,371 @@
+for creation.
+
+If only the fields or only the annotations are changed, take the new value and copy it to both the
+corresponding fields & annotations (even if one was previously unset).
+
+If both the fields _and_ annotations are changed, the new values MUST match. Enforced in API
+validation.
+
+#### PodSecurityPolicy Enforcement
+
+The PodSecurityPolicy admission controller must continue to check the PSP object for annotations if
+the fields are unset, in order to handle upgrade & version skew scenarios.
+
+When setting default profiles, PSP only needs to set the field. The API machinery will handle
+setting the annotation as necessary.
+
+When enforcing allowed profiles, the PSP should check BOTH the annotations & fields. In most cases,
+they should be consistent. On pod update, the seccomp annotations may differ from the fields. In
+that case, the PSP enforcement should check both values as the effective value depends on the node
+version running the pod.
+
+#### PodTemplates
+
+PodTemplates (e.g. ReplaceSets, Deployments, StatefulSets, etc.) will be ignored. The
+field/annotation resolution will happen on template instantiation.
+
+#### Upgrade / Downgrade
+
+Nodes do not currently support in-place upgrades, so pods will be recreated on node upgrade and
+downgrade. No special handling or consideration is needed to support this.
+
+On the API server side, we've already taken version skew in HA clusters into account. The same
+precautions make upgrade & downgrade handling a non-issue.
+
+### Test Plan
+
+Seccomp already has [E2E tests][], but the tests are guarded by the `[Feature:Seccomp]` tag and not
+run in the standard test suites.
+
+Prior to being marked GA, the feature tag will be removed from the seccomp tests, and the tests will
+be migrated to the new fields API. Tests will be tagged as `[LinuxOnly]`.
+
+New tests will be added covering the annotation/field conflict cases described under
+[Version Skew Strategy](#version-skew-strategy).
+
+Test coverage for localhost profiles will be added, unless we decide to [keep localhost support in
+alpha](#alternatives).
+
+[E2E tests]: https://github.com/kubernetes/kubernetes/blob/5db091dde4d7de74283ca94870958acf63010c0a/test/e2e/node/security_context.go#L147
+
+### Graduation Criteria
+
+_This section is excluded, as it is the subject of the entire proposal._
+
+### Upgrade / Downgrade Strategy
+
+See [Version Skew Strategy](#version-skew-strategy).
+
+## Implementation History
+
+- 2019-07-17: Initial KEP
+
+## Drawbacks
+
+Promoting seccomp as-is to GA may be seen as "blessing" the current functionality, and make it
+harder to make some of the enhancements listed under [Non-Goals](#non-goals). Since the current
+behavior is unguarded, I think we already need to treat the behavior as GA (which is why it's been
+so hard to change the default profile), so I do not think these changes will actually increase the
+friction.
```

Comment:
This is my biggest concern, but I think the arguments you've laid out are correct.
This is a necessary step to get to the good stuff :)

Thanks so much Tim!